### PR TITLE
Add support for sex chromosomes to per-sample summary stats

### PR DESCRIPTION
**Pull request by jkgoodrich**
#### Create filter group HT for all tests:
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --create-filter-group-ht \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
Filter group HT: gs://gnomad/v4.1/assessment/exomes/gnomad.exomes.v4.1.per_sample_filtering_groups.ht
```
Job ID: [55c5bbe6a3ff4abdab4581edf44dcad4](https://console.cloud.google.com/dataproc/jobs/55c5bbe6a3ff4abdab4581edf44dcad4/monitoring?region=us-central1&project=broad-mpg-gnomad)

#### No intermediate:

**Full pipeline with autosomes and sex chromosomes processed in one run:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-per-sample-counts-ht \
    --aggregate-sample-stats \
    --custom-suffix no_intermediate_all_in_one \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_all_in_one.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_all_in_one.per_sample_variant_counts.aggregated.ht
```
Job ID: [84e68a8a723045449d348e8dd131578c](https://console.cloud.google.com/dataproc/jobs/84e68a8a723045449d348e8dd131578c/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on autosomes only:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-per-sample-counts-ht \
    --aggregate-sample-stats \
    --autosomes-only-stats \
    --custom-suffix no_intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.autosomes.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.autosomes.aggregated.ht
```
Job ID: [e9e5619a71154b4b9ced0ae5b6f04bce](https://console.cloud.google.com/dataproc/jobs/e9e5619a71154b4b9ced0ae5b6f04bce/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on sex chromosomes only:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-per-sample-counts-ht \
    --aggregate-sample-stats \
    --sex-chr-only-stats \
    --custom-suffix no_intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.sex_chr.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.sex_chr.aggregated.ht
```
Job ID: [a04667e986224188a5cbfd6fdf8e3740](https://console.cloud.google.com/dataproc/jobs/a04667e986224188a5cbfd6fdf8e3740/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Combine autosomes and sex chromosomes stats and aggregate:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --aggregate-sample-stats \
    --combine-autosome-and-sex-chr-stats \
    --custom-suffix no_intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.aggregated.ht
```
Job ID: [b7c2e9eaea7549128cb30af010807f52](https://console.cloud.google.com/dataproc/jobs/b7c2e9eaea7549128cb30af010807f52/monitoring?region=us-central1&project=broad-mpg-gnomad)

#### Using intermediate:

**Full pipeline with autosomes and sex chromosomes processed in one run:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-intermediate-mt-for-sample-counts \
    --create-per-sample-counts-ht \
    --use-intermediate-mt-for-sample-counts \
    --aggregate-sample-stats \
    --custom-suffix intermediate_all_in_one \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
intermediate HT: gs://gnomad-tmp/gnomad.exomes.v4.1.qc_data/per_sample_summary_stats_intermediate.test.ht
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_all_in_one.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_all_in_one.per_sample_variant_counts.aggregated.ht
```
Job ID: [3ec7d2b378b24fdaaec691d483293585](https://console.cloud.google.com/dataproc/jobs/3ec7d2b378b24fdaaec691d483293585/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on autosomes only:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-intermediate-mt-for-sample-counts \
    --create-per-sample-counts-ht \
    --use-intermediate-mt-for-sample-counts \
    --aggregate-sample-stats \
    --autosomes-only-stats \
    --custom-suffix intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
intermediate HT: gs://gnomad-tmp/gnomad.exomes.v4.1.qc_data/per_sample_summary_stats_intermediate.test.autosomes.ht
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.autosomes.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.autosomes.aggregated.ht
```
Job ID: [7284b9db36de4909be6bcb6be6b9cb71](https://console.cloud.google.com/dataproc/jobs/7284b9db36de4909be6bcb6be6b9cb71/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on sex chromosomes only:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-intermediate-mt-for-sample-counts \
    --create-per-sample-counts-ht \
    --use-intermediate-mt-for-sample-counts \
    --aggregate-sample-stats \
    --sex-chr-only-stats \
    --custom-suffix intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
intermediate HT: gs://gnomad-tmp/gnomad.exomes.v4.1.qc_data/per_sample_summary_stats_intermediate.test.sex_chr.ht
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.sex_chr.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.sex_chr.aggregated.ht
```
Job ID: [dc1b64766e454f88897fe8d9b5dddc03](https://console.cloud.google.com/dataproc/jobs/dc1b64766e454f88897fe8d9b5dddc03/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Combine autosomes and sex chromosomes stats and aggregate:**
Command:
```
hailctl dataproc submit jg2 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --aggregate-sample-stats \
    --combine-autosome-and-sex-chr-stats \
    --custom-suffix intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.aggregated.ht
```
Job ID: [d5e4fb6a5a9445bba52e6f16dcd38ff4](https://console.cloud.google.com/dataproc/jobs/d5e4fb6a5a9445bba52e6f16dcd38ff4/monitoring?region=us-central1&project=broad-mpg-gnomad)

### CODE REVIEW

**Comment by mike-w-wilson** [comment]
This is a great rewrite, it is so clear. I totally forgot about the pipeline class and how much that cleans the script up -- as does the the test arg handling. As far as data out looks, the y_non_par stats are different when comparing the all in one run, (`gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_all_in_one.per_sample_variant_counts.aggregated.ht` from Job ID: [84e68a8a723045449d348e8dd131578c](https://console.cloud.google.com/dataproc/jobs/84e68a8a723045449d348e8dd131578c/monitoring?region=us-central1&project=broad-mpg-gnomad))  and the intermediate+combine run (`gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.aggregated.ht` from Job ID: [d5e4fb6a5a9445bba52e6f16dcd38ff4](https://console.cloud.google.com/dataproc/jobs/d5e4fb6a5a9445bba52e6f16dcd38ff4/monitoring?region=us-central1&project=broad-mpg-gnomad)).
Screenshot 2024-07-11 at 3 31 31 PM
The first column is a matching boolean, the second is the full run, the third is the intermediate+combine. The intermediate+combine appears to always be half the full run value so maybe something is happening regarding ploidy in the full? Here is my notebook (gs://gnomad-mwilson/v4.1/sum_stat_testing.ipynb), last cell is the fails. Other than that, this looks good, all my comments are for comments/docs.

**Comment by jkgoodrich** [comment]
OK, I found the problem, it's a very simple fix. Just needed to move the setting of y_nonpar in XX samples to missing so that it's also done on the combined HT.

### Here are the new tests:

#### No intermediate:

**Full pipeline with autosomes and sex chromosomes processed in one run:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-per-sample-counts-ht \
    --aggregate-sample-stats \
    --custom-suffix no_intermediate_all_in_one \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_all_in_one.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_all_in_one.per_sample_variant_counts.aggregated.ht
```
Job ID: [cf9d6f6773d0466789324c7512fea122](https://console.cloud.google.com/dataproc/jobs/cf9d6f6773d0466789324c7512fea122/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on autosomes only:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-per-sample-counts-ht \
    --aggregate-sample-stats \
    --autosomes-only-stats \
    --custom-suffix no_intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.autosomes.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.autosomes.aggregated.ht
```
Job ID: [1d3c35cad8704f85be6927734d19e565](https://console.cloud.google.com/dataproc/jobs/1d3c35cad8704f85be6927734d19e565/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on sex chromosomes only:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-per-sample-counts-ht \
    --aggregate-sample-stats \
    --sex-chr-only-stats \
    --custom-suffix no_intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.sex_chr.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.sex_chr.aggregated.ht
```
Job ID: [41a2325bf5804cd2ac6277b9ae6be5d6](https://console.cloud.google.com/dataproc/jobs/41a2325bf5804cd2ac6277b9ae6be5d6/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Combine autosomes and sex chromosomes stats and aggregate:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --aggregate-sample-stats \
    --combine-autosome-and-sex-chr-stats \
    --custom-suffix no_intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.no_intermediate_combine_chr.per_sample_variant_counts.aggregated.ht
```
Job ID: [b012882d71894e3ba4fd7aac02bf225d](https://console.cloud.google.com/dataproc/jobs/b012882d71894e3ba4fd7aac02bf225d/monitoring?region=us-central1&project=broad-mpg-gnomad)



#### Using intermediate:

**Full pipeline with autosomes and sex chromosomes processed in one run:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-intermediate-mt-for-sample-counts \
    --create-per-sample-counts-ht \
    --use-intermediate-mt-for-sample-counts \
    --aggregate-sample-stats \
    --custom-suffix intermediate_all_in_one \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
intermediate HT: gs://gnomad-tmp/gnomad.exomes.v4.1.qc_data/per_sample_summary_stats_intermediate.test.ht
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_all_in_one.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_all_in_one.per_sample_variant_counts.aggregated.ht
```
Job ID: [ed4da37139e04ba09b8eb34f163a0693](https://console.cloud.google.com/dataproc/jobs/ed4da37139e04ba09b8eb34f163a0693/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on autosomes only:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-intermediate-mt-for-sample-counts \
    --create-per-sample-counts-ht \
    --use-intermediate-mt-for-sample-counts \
    --aggregate-sample-stats \
    --autosomes-only-stats \
    --custom-suffix intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
intermediate HT: gs://gnomad-tmp/gnomad.exomes.v4.1.qc_data/per_sample_summary_stats_intermediate.test.autosomes.ht
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.autosomes.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.autosomes.aggregated.ht
```
Job ID: [b85117e5f62f434ca30480ccf3b3f94f](https://console.cloud.google.com/dataproc/jobs/b85117e5f62f434ca30480ccf3b3f94fmonitoring?region=us-central1&project=broad-mpg-gnomad)

**Full pipeline on sex chromosomes only:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --create-intermediate-mt-for-sample-counts \
    --create-per-sample-counts-ht \
    --use-intermediate-mt-for-sample-counts \
    --aggregate-sample-stats \
    --sex-chr-only-stats \
    --custom-suffix intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
intermediate HT: gs://gnomad-tmp/gnomad.exomes.v4.1.qc_data/per_sample_summary_stats_intermediate.test.sex_chr.ht
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.sex_chr.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.sex_chr.aggregated.ht
```
Job ID: [65d226b70bdd4ca6baa279e70dd6a841](https://console.cloud.google.com/dataproc/jobs/65d226b70bdd4ca6baa279e70dd6a841/monitoring?region=us-central1&project=broad-mpg-gnomad)

**Combine autosomes and sex chromosomes stats and aggregate:**
Command:
```
hailctl dataproc submit jg1 gnomad_qc/v4/assessment/calculate_per_sample_stats.py \
    --data-type exomes \
    --test-dataset \
    --aggregate-sample-stats \
    --combine-autosome-and-sex-chr-stats \
    --custom-suffix intermediate_combine_chr \
    --overwrite \
    --pyfiles ../gnomad_methods/gnomad,gnomad_qc
```
Output: 
```
per-sample counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.ht
aggregate counts: gs://gnomad-tmp/gnomad_v4.1_testing/assessment/exomes/gnomad.exomes.v4.1.intermediate_combine_chr.per_sample_variant_counts.aggregated.ht
```
Job ID: [00fe80d228bf4682a3fa0a0ba5b58065](https://console.cloud.google.com/dataproc/jobs/00fe80d228bf4682a3fa0a0ba5b58065/monitoring?region=us-central1&project=broad-mpg-gnomad)


Updated the notebooks:
```
http://localhost:8123/notebooks/gnomad-julia/gnomad_v4/test_per_sample_stats_add_sex_chr.ipynb#
http://localhost:8123/notebooks/gnomad-mwilson/v4.1/sum_stat_testing.ipynb#
```

File: gnomad_qc/v4/assessment/calculate_per_sample_stats.py [hunk 1]
```diff
@@ -390,6 +587,115 @@ def prepare_mt_for_sample_counts(
...
+        Default is {}.
+    :return: Table containing per-sample stats annotated with the requested ratios.
+    """
+    return ht.annotate(
+        summary_stats=ht.summary_stats.map(
+            lambda x: x.annotate(
+                **{k: hl.sum([x[s] for s in v]) for k, v in sums.items()},
+                **{k: x[d1] - x[d2] for k, (d1, d2) in diffs.items()},
+                **{
+                    k: divide_null(hl.float64(x[n]), x[d])
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/assessment/calculate_per_sample_stats.py:636]: Can you take a look again? I only see NaNs now in the agg Tables, not the count Tables

**Review by mike-w-wilson — APPROVED** [review]
:shipit:

**Pull request merged**